## The purpose of the study

Based on the data provided, form a value proposition for a new product: **App for book lovers**

**Содержание:**

[1. Overview of tables](#paragraph1)  
[2. Counting the number of books](#paragraph2)  
[3. Counting the number of reviews](#paragraph3)  
[4. Definition of publishing house](#paragraph4)  
[5. Definition of the author](#paragraph5)  
[6. Counting the number of reviews](#paragraph6)  
[7. Output](#paragraph7)


importing libraries:

In [1]:
import pandas as pd
from sqlalchemy import create_engine

Connecting to the database:

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### 1. Overview of tables<a name="paragraph1"></a>

In [3]:
pd.io.sql.read_sql('SELECT * FROM books LIMIT 10', con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [4]:
pd.io.sql.read_sql('SELECT * FROM authors LIMIT 10', con = engine)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [5]:
pd.io.sql.read_sql('SELECT * FROM publishers LIMIT 10', con = engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [6]:
pd.io.sql.read_sql('SELECT * FROM ratings LIMIT 10', con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [7]:
pd.io.sql.read_sql('SELECT * FROM reviews LIMIT 10', con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

### 2. Counting the number of books released since January 2000<a name="paragraph2"></a>

In [8]:
query = '''
SELECT
    COUNT(book_id) as book_cnt
FROM
    books
WHERE
    publication_date > '2000-01-02'
    '''

In [9]:
pd.io.sql.read_sql(query, con = engine)

book_cnt
0       819

There were 819 books in the database

### 3. Counting the number of user reviews and the average rating of each book<a name="paragraph3"></a>

In [10]:
query = '''
SELECT
    books.book_id,
    books.title,
    COUNT(DISTINCT review_id) as review_cnt,
    AVG(rating) as avg_rating
FROM 
    books
INNER JOIN reviews ON  reviews.book_id = books.book_id
INNER JOIN ratings ON  ratings.book_id = books.book_id
GROUP BY
    books.book_id
ORDER BY
    review_cnt DESC
LIMIT 7
'''

In [11]:
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  review_cnt  \
0      948                            Twilight (Twilight  #1)           7   
1      656                                     The Book Thief           6   
2      207                                    Eat  Pray  Love           6   
3      299  Harry Potter and the Chamber of Secrets (Harry...           6   
4      497                          Outlander (Outlander  #1)           6   
5      627                                      The Alchemist           6   
6      302  Harry Potter and the Prisoner of Azkaban (Harr...           6   

   avg_rating  
0    3.662500  
1    4.264151  
2    3.395833  
3    4.287500  
4    4.125000  
5    3.789474  
6    4.414634

In [12]:
#pd.io.sql.read_sql(query, con = engine).review_cnt.sum()

Most user reviews were written for the book "Twilight" - 7

### 4. Let's find out the publishing house that has released the largest number of books thicker than 50 pages<a name="paragraph4"></a>

In [13]:
query = '''
SELECT
    publisher,
    COUNT(book_id) as book_cnt
FROM
    publishers
LEFT JOIN books ON  books.publisher_id = publishers.publisher_id 
WHERE 
    num_pages > 50
GROUP BY
    publisher
ORDER BY
    book_cnt DESC
'''

In [14]:
pd.io.sql.read_sql(query, con = engine)

publisher  book_cnt
0                                        Penguin Books        42
1                                              Vintage        31
2                             Grand Central Publishing        25
3                                     Penguin Classics        24
4                                     Ballantine Books        19
..                                                 ...       ...
329                                         Turtleback         1
330  Atheneum Books for Young Readers: Richard Jack...         1
331                                     Penguin Signet         1
332                                    Victor Gollancz         1
333                      Harvard Business Review Press         1

[334 rows x 2 columns]

Most of the books have been published by the publishing house **Penguin Books**

### 5. Find out the author with the highest average rating of his works (of those with more than 50 ratings)<a name="paragraph5"></a>

In [15]:
query = '''
SELECT
    author,
    AVG(sub.avg_rating) as total_avg_rating
FROM
    (SELECT
        author,
        books.book_id,
        AVG(rating) as avg_rating
    FROM 
        authors
    LEFT JOIN books ON books.author_id = authors.author_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
    GROUP BY
        author,
        books.book_id
    HAVING
        COUNT(rating) > 50) as sub
GROUP BY
    author
ORDER BY
    total_avg_rating DESC
'''

In [16]:
pd.io.sql.read_sql(query, con = engine)

author  total_avg_rating
0                          J.K. Rowling/Mary GrandPré          4.283844
1                   Markus Zusak/Cao Xuân Việt Khương          4.264151
2                                      J.R.R. Tolkien          4.258446
3                                   Louisa May Alcott          4.192308
4                                        Rick Riordan          4.080645
5                                     William Golding          3.901408
6                                       J.D. Salinger          3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce          3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...          3.787879
9                                           Dan Brown          3.754540
10                                         Lois Lowry          3.750000
11            George Orwell/Boris Grabnar/Peter Škerl          3.729730
12                                    Stephenie Meyer          3.662500
13                                     John Steinbeck          3.622951

**J.K. Rowling has the highest marks**

### 6. The average number of text reviews among those readers who gave more than 50 ratings to books<a name="paragraph6"></a>

In [17]:
query = '''
SELECT
    ROUND(AVG(sub_text_cnt.text_cnt)) AS avg_text_cnt
FROM
    (SELECT
        COUNT(text) as text_cnt
    FROM 
        reviews
    WHERE username IN
        (SELECT username
        FROM 
            ratings
        GROUP BY
            username
        HAVING
            COUNT(ratings) > 50)
    GROUP BY 
            username
    ) AS sub_text_cnt  
'''

In [18]:
pd.io.sql.read_sql(query, con = engine)

avg_text_cnt
0          24.0

The average number of reviews written by readers who have rated more than 50 books is 24

### 7. Conclusions:<a name="paragraph7"></a>

Based on the conducted research, we can say that:  
- The most reviews have been written for the book **tWilight **  
- The highest rating for the book **Harry Potter and the Prisoner of Azkaban**  
- Publisher **Penguin Books** has released the most books thicker than 50 pages  
- **J.K. Rowling** has the highest average score among authors with more than 50 ratings  
- Users who have given more than 50 ratings to books, on average write ** 24 reviews**  